src/libraries/Native/AnyOS/System.Globalization.Native/pal_calendarData.c

```diff
@@ -306,7 +313,11 @@ static int32_t EnumSymbols(const char* locale,
         return FALSE;
 
     char localeWithCalendarName[ULOC_FULLNAME_CAPACITY];
+#ifdef TARGET_UNIX
     strncpy(localeWithCalendarName, locale, ULOC_FULLNAME_CAPACITY);
+#else
+    strncpy_s(localeWithCalendarName, sizeof(locale), locale, ULOC_FULLNAME_CAPACITY);
```

Comment:
This is lame. Can we just shut off the Windows warnings that are preventing perfectly fine code from getting compiled?

Reply:
Also, the Windows version of the code has a (potentially?) bug.

Reply:
> Can we just shut off the Windows warnings that are preventing perfectly fine code from getting compiled?

Or make a helper function so that we do not need to ifdef every use.

Reply:
maybe something like:

```diff
--- a/src/libraries/Native/AnyOS/System.Globalization.Native/pal_calendarData.c
+++ b/src/libraries/Native/AnyOS/System.Globalization.Native/pal_calendarData.c
@@ -28,6 +28,12 @@
 #define ISLAMIC_UMALQURA_NAME "islamic-umalqura"
 #define ROC_NAME "roc"
 
+#ifdef TARGET_UNIX
+#define STRING_COPY(localeNamePtr, size, locale, capacity) strncpy(localeNamePtr, locale, capacity);
+#else
+#define STRING_COPY(localeNamePtr, size, locale, capacity) strncpy_s(localeNamePtr, size, locale, capacity);
+#endif
+
 #define JAPANESE_LOCALE_AND_CALENDAR "ja_JP@calendar=japanese"
 
 static const UChar UDAT_MONTH_DAY_UCHAR[] = {'M', 'M', 'M', 'M', 'd', '\0'};
@@ -313,11 +319,7 @@ static int32_t EnumSymbols(const char* locale,
         return FALSE;
 
     char localeWithCalendarName[ULOC_FULLNAME_CAPACITY];
-#ifdef TARGET_UNIX
-    strncpy(localeWithCalendarName, locale, ULOC_FULLNAME_CAPACITY);
-#else
-    strncpy_s(localeWithCalendarName, sizeof(locale), locale, ULOC_FULLNAME_CAPACITY);
-#endif
+    STRING_COPY(localeWithCalendarName, sizeof(locale), locale, ULOC_FULLNAME_CAPACITY);
     localeWithCalendarName[ULOC_FULLNAME_CAPACITY - 1] = 0;
 
     uloc_setKeywordValue("calendar", GetCalendarName(calendarId), localeWithCalendarName, ULOC_FULLNAME_CAPACITY, &err);
@@ -425,11 +427,7 @@ static int32_t EnumAbbrevEraNames(const char* locale,
 
     char* localeNamePtr = localeNameBuf;
     char* parentNamePtr = parentNameBuf;
-#ifdef TARGET_UNIX
-    strncpy(localeNamePtr, locale, ULOC_FULLNAME_CAPACITY);
-#else
-    strncpy_s(localeNamePtr, sizeof(locale), locale, ULOC_FULLNAME_CAPACITY);
-#endif
+    STRING_COPY(localeNamePtr, sizeof(locale), locale, ULOC_FULLNAME_CAPACITY);
     localeNamePtr[ULOC_FULLNAME_CAPACITY - 1] = 0;
 
     while (TRUE)
```

Reply:
The zero termination should be part of the thing. `strncpy_s` does it for you, so no need to repeat it.

Reply:
True, updated. I was thinking that `strncpy_s` would just work on Unix, but since we are std=gnu99 compliant here, `strncpy_s` is not available (at least on macOS).

Reply:
why on Linux we need do-while(0)?

Reply:
Note that you should be able to suppress the warnings on Windows about `strncpy` with `_CRT_SECURE_NO_DEPRECATE`. (This is what the ICU source does.)

Reply:
It may cheaper to have the helper function with ifdefs. strncpy will get flagged by the Windows static analysis tools, and so we will need to justify that it is really ok every once in a while.

Reply:
You might need to be careful with the [constraint handler](https://en.cppreference.com/w/c/string/byte/strncpy) for `strncpy_s`.
Ex: Somewhere else might set it to call `abort()` on truncation.
Maybe that is a good thing though, but I'm not sure what you want for your code.

Reply:
Right, that is the bug that I have mentioned above.

Reply:
@tarekgh, in C/C++, to have multi-line macro, `do-while(0)` approach is used (multiple places in CoreCLR itself; not platform specific).

Reply:
thanks @am11